053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,465 @@
+  when Concourse starts up. For an `Elasticsearch` implementation of
+  `EventStore`, for instance, `Setup()` would likely initialize a connection to
+  the Elasticsearch cluster. For a `Postgres` implementation, `Setup()` will
+  check if the `build_events` table exists, and if not, create it (note: this
+  was previously done by DB migrations - more on this in [Migrations](#migrations))
+
+* `Close(ctx)` is for cleaning up any left over resources. For instance, if a
+  connection was opened to an Elasticsearch cluster in `Setup`, `Close` should
+  gracefully terminate the connection. However, it should not necessarily be the
+  inverse of `Setup` - e.g. for `Postgres`, `Setup` creates the `build_events`
+  table, but `Close` should not drop it.
+
+* `Initialize(ctx, build db.Build)` will be triggered when a build is first created, and
+  before any build events are `Put` into the store. For the `Postgres`
+  implementation, this will create a sequence like `build_event_id_seq_x` for the
+  current build. It will also create either the table `pipeline_build_events_x`
+  or `team_build_events_x`, depending on whether it's a one-off build or not.
+  This represents a behaviour change from what we have now, where this table is
+  created when the team/pipeline is first created, but it shouldn't really have
+  much impact.
+
+* `Finalize(ctx, build db.Build)` will be triggered when a build is `Finish`ed. No
+  build events may be `Put` into the store after this is called. For
+  `Postgres`, this involves dropping the `build_event_id_seq_x` sequence.
+
+* `Put(ctx, build db.Build, events []atc.Event)` is called whenever
+  `build.SaveEvent(event)` is called in order to write build events to the
+  external store. The reason for accepting a list of `events` (despite
+  `db.Build.SaveEvent` only ever passing in a single event) is because if we
+  want to migrate build events to another `EventStore`, it would be super slow
+  making a huge number of small `Put` calls when we can do one batch `Put` per
+  Build.
+
+* `Get(ctx, build db.Build, requested int, cursor *Key)` fetches events from the
+  `EventStore`, starting from an initial `Key` (which is excluded from the
+  result set - e.g.  if `cursor` points to the 10th `Key`, the 11th event will be
+  the next one fetched). `Key` can be any type, and is implementation-specific.
+  For `Postgres`, this will likely be something like `uint`, but other stores
+  (e.g. DynamoDB) that don't provide easy support for auto-incrementing IDs may
+  choose a different type of `Key`. Regardless of the type of `Key`, if the goal
+  is to fetch all events, `cursor` should initially point to `nil` (but the
+  pointer itself must not be `nil`). For instance, you can iterate over all
+  events for a completed build like so:
+
+  ```go
+  var cursor Key
+  batchSize := 1000
+  for {
+      events, err := eventStore.Get(build, batchSize, &cursor)
```

Comment:
minor nitpick: I noticed the `ctx context.Context` is added to the interface functions recently. It is missed in the param list here and other embedded code snippets.